neonvm/controllers/vm_controller_unit_test.go

```diff
@@ -0,0 +1,262 @@
+
+type testParams struct {
+	t            *testing.T
+	ctx          context.Context
+	r            *VMReconciler
+	client       client.Client
+	origVM       *vmv1.VirtualMachine
+	mockRecorder *mockRecorder
+}
+
+var reconcilerMetrics = MakeReconcilerMetrics()
+
+func newTestParams(t *testing.T) *testParams {
+	os.Setenv("VM_RUNNER_IMAGE", "vm-runner-img")
+
+	logger := zap.New(zap.UseDevMode(true), zap.WriteTo(os.Stdout),
+		zap.Level(zapcore.DebugLevel))
+	ctx := log.IntoContext(context.Background(), logger)
+
+	scheme := runtime.NewScheme()
+	scheme.AddKnownTypes(vmv1.SchemeGroupVersion, &vmv1.VirtualMachine{})
+	scheme.AddKnownTypes(corev1.SchemeGroupVersion, &corev1.Pod{})
+
+	params := &testParams{
+		t:      t,
+		ctx:    ctx,
+		client: fake.NewClientBuilder().WithScheme(scheme).Build(),
+		//nolint:exhaustruct // This is a mock
+		mockRecorder: &mockRecorder{},
+		r:            nil,
+		origVM:       nil,
+	}
+
+	params.r = &VMReconciler{
+		Client:   params.client,
+		Recorder: params.mockRecorder,
+		Scheme:   scheme,
+		Config: &ReconcilerConfig{
+			IsK3s:                   false,
+			UseContainerMgr:         false,
+			MaxConcurrentReconciles: 10,
+			QEMUDiskCacheSettings:   "",
+			FailurePendingPeriod:    time.Minute,
+			FailingRefreshInterval:  time.Minute,
+		},
+		Metrics: reconcilerMetrics,
+	}
+
+	return params
+}
+
+func (p *testParams) initVM(vm *vmv1.VirtualMachine) *vmv1.VirtualMachine {
+	err := p.client.Create(p.ctx, vm)
+	require.NoError(p.t, err)
+	p.origVM = vm
+
+	// Do serialize/deserialize, to normalize resource.Quantity
+	return p.getVM()
+}
+
+func (p *testParams) getVM() *vmv1.VirtualMachine {
+	var obj vmv1.VirtualMachine
+	err := p.client.Get(p.ctx, client.ObjectKeyFromObject(p.origVM), &obj)
+	require.NoError(p.t, err)
+
+	return &obj
+}
+
+func TestReconcile(t *testing.T) {
+	params := newTestParams(t)
+	origVM := params.initVM(defaultVm())
+
+	req := reconcile.Request{
+		NamespacedName: client.ObjectKeyFromObject(origVM),
+	}
+
+	// Round 1
+	res, err := params.r.Reconcile(params.ctx, req)
+	assert.NoError(t, err)
+
+	// Added finalizer
+	assert.Equal(t, reconcile.Result{
+		Requeue: true,
+	}, res)
+	assert.Contains(t, params.getVM().Finalizers, virtualmachineFinalizer)
+
+	// Round 2
+	res, err = params.r.Reconcile(params.ctx, req)
+	assert.NoError(t, err)
+	assert.Equal(t, false, res.Requeue)
+
+	// VM is pending
+	assert.Equal(t, vmv1.VmPending, params.getVM().Status.Phase)
+
+	// Round 3
+	params.mockRecorder.On("Event", mock.Anything, "Normal", "Created",
+		mock.Anything)
+	res, err = params.r.Reconcile(params.ctx, req)
+	assert.NoError(t, err)
+	assert.Equal(t, false, res.Requeue)
+
+	// We now have a pod
+	vm := params.getVM()
+	assert.NotEmpty(t, vm.Status.PodName)
+	// Spec is unchanged
+	assert.Equal(t, vm.Spec, origVM.Spec)
+
+	// Round 4
+	res, err = params.r.Reconcile(params.ctx, req)
+	assert.NoError(t, err)
+	assert.Equal(t, false, res.Requeue)
+
+	// Pod is pending, so nothing changes
```

Comment:
Strictly speaking, nothing is moving the pod along, so it'll still have a phase of `""`, right? (or, do you mean "pending" in the generic sense here?)

Reply:
Perfect, nice